src/mlpack/methods/rbm/ssRBM.hpp

```diff
@@ -0,0 +1,282 @@
+
+  // Variable for Positive grad wrt weights
+  weightPositiveGrad = arma::cube(positiveGradient.memptr(),
+    visible.PoolSize(),
+      visible.InSize(),
+      visible.OutSize(),
+    false, false);
+
+  // Variable for Positive grad wrt hidden bias
+  hiddenBiasPositiveGrad = arma::mat(
+    positiveGradient.memptr() + weightPositiveGrad.n_elem,
+    1, numHidden, false, false);
+
+  // Variable for Positive grad wrt visible bias
+  visibleBiasPositiveGrad = arma::mat(
+      positiveGradient.memptr() + weightPositiveGrad.n_elem +
+      hiddenBiasPositiveGrad.n_elem,
+      numVisible, 1, false, false);
+  }
+
+  void ResetGradient()
+  {
+    weightPositiveGrad.set_size(visible.Weight().n_rows,
+        visible.Weight().n_cols,
+        visible.Weight().n_slices);
+
+    weightNegativeGrad.set_size(visible.Weight().n_rows,
+        visible.Weight().n_cols,
+        visible.Weight().n_slices);
+
+    visibleBiasPositiveGrad.set_size(numVisible, 1);
+    visibleBiasNegativeGrad.set_size(numVisible, 1);
+
+    hiddenBiasPositiveGrad.set_size(1, numHidden);
+    hiddenBiasNegativeGrad.set_size(1, numHidden);
+  }
+
+  /**
+   * Free energy of the spike and slab variable
+   * the free energy of the ssRBM is given my
+   * $v^t$$\Delta$v - $\sum_{i=1}^N$ 
+   * $\log{ \sqrt{\frac{(-2\pi)^K}{\prod_{m=1}^{K}(\alpha_i)_m}}}$ -
+   * $\sum_{i=1}^N \log(1+\exp( b_i +
+   * \sum_{m=1}^k \frac{(v(w_i)_m^t)^2}{2(\alpha_i)_m})$
+   *
+   * @param input the visible layer
+   */ 
+  double FreeEnergy(arma::mat&& input)
+  {
+    const double pi = boost::math::constants::pi<double>();
+    freeEnergySum = 0;
+    double temp = arma::as_scalar(input.t() *
+        arma::diagmat(visible.LambdaBias()) * input);
+    for (size_t i = 0; i < numHidden; i++)
+    {
+      temp = std::log(std::sqrt(std::pow(-2 * pi, 2) /
+          arma::as_scalar(arma::prod(hidden.SlabBias().col(i)))));
+    }
+
+    for (size_t i = 0; i < numHidden; i++)
+    {
+      for (size_t j = 0; j < hidden.SlabBias().n_cols; j++)
+        freeEnergySum +=
+            SoftplusFunction::Fn(arma::as_scalar(hidden.SpikeBias().col(i)) +
+            std::pow(arma::as_scalar(visible.Weight().slice(i).row(j) *
+                input), 2) / arma::as_scalar(hidden.SlabBias().col(i).row(j)));
+    }
+
+    return temp + freeEnergySum;
+  }
+
+  /**
+   * Gradient function calculates the gradient for the spike and
+   * slab RBM.
+   *
+   * @param input the visible input
+   * @param output the computed gradient
+   */
+  void PositivePhase(arma::mat&& input)
+  {
+    positiveGradient.zeros();
+    // positive phase
+    visible.Sample(std::move(input), std::move(tempHidden));
+
+    visible.Psgivenvh(std::move(input), std::move(tempHidden),
+        std::move(tempMean), std::move(tempSlab));
+
+    // positive weight gradient
+    for (size_t i = 0 ; i < tempHidden.n_rows; i++)
+    {
+      weightPositiveGrad.slice(i) = tempMean * input.t() *
+          arma::as_scalar(tempHidden.row(i));
```

Comment:
If I am not mistaken `arma::as_scalar(tempHidden.row(i))` is equal to `tempHidden[i]` since actually `tempHidden` is a vector.